prediction/src/preprocess/improved_lung_segmentation.py

```diff
@@ -0,0 +1,444 @@
+
+    To determine the upper threshold for the
+    region growing operation, optimal thresholding is applied.
+
+    Args:
+        patient (np.ndarray[bool]): a label of CT
+
+    Returns:
+        np.ndarray[bool]: a label of the segmented lungs.
+
+    """
+    lungs_seeds = patient * seeds == patient[seeds].min()
+    lungs_seeds = lungs_seeds.astype(np.bool_)
+    threshold = skimage.filters.threshold_otsu(patient[patient.shape[0] // 2])
+
+    lungs_seeds = region_growing(patient.copy(), lungs_seeds, -1010, threshold)
+    return scipy.ndimage.morphology.binary_opening(lungs_seeds - scipy.ndimage.morphology.binary_opening(seeds))
+
+
+def lungs_postprocessing(lungs_seeds):
+    """ Smoothing.
+
+    First 3D hole filling is applied
+    to include vessels and other high-density structures that were
+    excluded by the threshold used in region growing, in the segmentation.
+
+    Next, morphological closing with a spherical
+    kernel is applied. The diameter of the kernel is set to 2% of
+    the x-dimension of the image.
+
+    Args:
+        lungs_seeds (np.ndarray[bool]): a label with the segmented lungs before postprocessing
+
+    Returns:
+        np.ndarray[bool]: a label with the segmented lungs after binary_fill_holes
+
+    """
+    for i in range(lungs_seeds.shape[1]):
+        lungs_seeds[:, i] = scipy.ndimage.morphology.binary_fill_holes(lungs_seeds[:, i])
+
+    return lungs_seeds
+
+
+def flip_lung(patient, trachea):
+    """ Аlipping of the lungs.
+
+    If no suitable region is detected  in the top slices,
+    the bottom slices of the scan are inspected  to be able
+    to handle cases that were scanned in a reverse  direction.
+
+    Args:
+        patient(np.ndarray[bool]): a label of CT
+        trachea (np.ndarray[bool]): a label of the trachea
+
+    Returns:
+        np.ndarray[bool]: patient - converted Lungs
+    """
+    z_coord, y_coord, x_coord = np.where(trachea)
+    z, y, x = np.where(patient)
+
+    z = np.unique(z)
+    z_mean = np.mean(np.unique(z_coord), axis=0)
+    z_down = abs(z[0] - z_mean)
+    z_up = abs(z[-1] - z_mean)
+    if z_down > z_up:
+        patient = np.flip(patient, axis=0)
+    return patient
+
+
+def conventional_lung_segmentation(patient):
+    """ The lungs segmentation.
+
+    The algorithm consists of the following steps:
+        (1)Extraction of large airways;
+        (2)Segmentation of lung regions;
+        (3)Separation of the left and right lungs;
+        (4)Smoothing.
+
+    The function aggregates extract_seeds, growing_bronchis,
+    flip_lung, grow_lungs, lungs_postprocessing.
+
+    Args:
+        patient (np.ndarray[bool]): a label of CT
+
+    Returns:
+        np.ndarray[bool]: lungs_seeds - a label with the segmented lungs.
+        np.ndarray[bool]: trachea - a label with the segmented trachea.
+    """
+
+    seeds = extract_seeds(patient)
+    trachea, _ = growing_bronchis(patient, seeds)
+#     patient = flip_lung(patient, trachea)
```

Comment:
I guess the commented code can be removed?

Reply:
Yes, you are absolutely right